### ISSUE TITLE
New way to document directives

### ISSUE THREAD

**Issue by olivernybroe** [nuwave/lighthouse#785]
Right now we have a problem in the way directives are documented. When going to a directive there is no way to see the arguments without reading the code and understanding it, which in some cases are easy and other cases aren't.

I am proposing a new way for us to document the directive from inside the php class. I was thinking we should use a phpdoc directive to have the SDL of the directive stored. SO for example we use a new phpdoc directive called SDL and then store the information of the SDL for the directive in here. Doing this will give us the advantage of easily knowing what arguments are available in the directive and also possibilities for auto generation as we can just combine all of the phpdocs from the directives to get a full SDL for all of the directives.

As this change will only affect the phpdocs, this will also be a non breaking change.

#782

**Comment by alberthaff** [comment]
We should be able to simply copy-paste the SDL once #784 is ready.

**Comment by spawnia** [comment]
I think we can just change the `Directive` interface. The `name` method is pretty pointless anyway, since we depend on the naming convention of the class name.

We can release this in `v4`, will be one more compelling reason to make a switch.

```php
<?php

namespace Nuwave\Lighthouse\Support\Contracts;

interface Directive
{
    /**
     * Formal definition of the directive using SDL.
     *
     * @return string
     */
    public function definition();
}
```

**Comment by olivernybroe** [comment]
@spawnia Just out of curiosity, what is the advantage of having it as a interface compared to having it part of the phpdoc? 
Only thing I can see is that we can force the user to add the documentation.

**Comment by spawnia** [comment]
Many.

PHPDocs require a special parser which leads to a more complex implementation from our side.
We might require the definition at runtime (e.g. for using the default args as defined in there), which would be slower and more error prone.

There is little autocompletion/IDE help for PHPDocs, especially custom annotations. With plain PHP, we can do something like this:

![image](https://user-images.githubusercontent.com/12158000/57635325-31fa8280-75a7-11e9-82de-474ab3ef8991.png)

I think the definition is slim enough and the benefits are plenty, that warrants forcing a definition from the user. It opens up new possibilites and niceties.

> Oh I guess one advantage is args can be dynamic as definition method is not static also?

True, we even have a use case for that: The `maxCount` argument of the `@paginate` defaults to the value set in the config.

**Comment by alberthaff** [comment]
> Oh I guess one advantage is args can be dynamic as definition method is not static also?

That would be a pretty cool feature. I did not think about that.

**Comment by olivernybroe** [comment]
Should we close this issue :) Sounds like we have a good way of doing this in another issue.
